src/config.ts

```diff
@@ -1,11 +1,15 @@
+import { promises as fs } from 'fs'
+import { basename, extname, dirname, join } from 'path'
+
 import mergeOptions from 'merge-options'
 
+import type { FeatureFlags } from './feature_flags.js'
 import { FunctionSource } from './function.js'
 import type { NodeBundlerName } from './runtimes/node/bundlers/types.js'
 import type { NodeVersionString } from './runtimes/node/index.js'
 import { minimatch } from './utils/matching.js'
 
-export interface FunctionConfig {
+interface FunctionConfig {
   externalNodeModules?: string[]
   includedFiles?: string[]
   includedFilesBasePath?: string
```

Comment:
We need one more field here: `nodeBundler?: string` to allow for bundler customisation.

Reply:
My bad, `nodeBundler` format needs to take a path to the main file as well as the format for file configuration. i.e `nodeBundler?: NodeBundlerConfig`.

I'm also now thinking that this should come with work on allowing of skipping bundling altogether: https://github.com/netlify/pillar-runtime/issues/201

```ts
interface NodeBundlerConfig {
    mainFile: string,
    moduleFormat: string
}
```

----------

example json
```json
{
  "nodeBundler": {
    "mainFile": "/some/path/to/a/function.js",
    "moduleFormat": "esm"
  },
  "includedFiles": ["some/path/**/*.js", "!other/path/**/*.js"]
}
```

Reply:
We may also not need to specify mainFile here since we are already handling one function at a time at this point, we may be able to get away with: `nodeBundler?: NodeBundlerName`.

Reply:
I think you still need a `mainFile` field, because we need to know which file is the entry point to the Lambda.

Reply:
But we are having the convention here that `function.js` has the config file `function.json` ins the same folder. So wouldn't function.js always be the main file?

Reply:
`function.js` is a source file, whereas `mainFile` is the distribution/compiled file that lives in the ZIP archive. Sometimes those two files will be the same, but not always. For example, when you're dealing with a TypeScript function, the source file will be something like `function.ts`, but your `mainFile` has to be a JavaScript because the Lambda runtime doesn't understand TypeScript.

You can see an example of that [here](https://github.com/netlify/zip-it-and-ship-it/blob/main/src/runtimes/node/bundlers/esbuild/index.ts#L112)

Reply:
Also for the nideBundler:  I think it makes sense to make it more clear to "users" what the config actually does if it is just that:

```json
"nodeBundler": {
  "mainFile": "/some/path/to/a/function.js",
  "moduleFormat": "esm"
},
```
I would not know what this actually does.

I think it would be more clear if we would for all bundlers do something like this:

```ts
type NodeBundlerName = 'esbuild' | 'nft' | 'zisi';
interface NodeBundler<N extends NodeBundlerName | 'none', O = never> {
  name: N
  options: O
}

interface UnbundledOptions {
  "mainFile":string
  "moduleFormat": "esm" | 'commonjs'
}

interface Config {
  nodeBundler: NodeBundlerName | NodeBundler<NodeBundlerName> | NodeBundler<'none', UnbundledOptions>
  ...
}
```

Allowing direct names as a convenience for all current bundlers, as they do not have options (yet?). `nodeBundler: 'esbuild'`. But it could also be written like `nodeBundler: { name: 'esbuild' }`. The exception is the `none` bundler which has some mandatory options, so no shortcut is allowed.
This way by looking at the config or its types it is always clear which bundler is configured.

Reply:
> `function.js` is a source file, whereas `mainFile` is the distribution/compiled file that lives in the ZIP archive. Sometimes those two files will be the same, but not always. For example, when you're dealing with a TypeScript function, the source file will be something like `function.ts`, but your `mainFile` has to be a JavaScript because the Lambda runtime doesn't understand TypeScript.

I thought if bundling is disabled, we would expect `functions.js` to be the already pre-bundled version. In this case we do not need the source of the function or do we?

Reply:
Hmm good point, maybe we don't need it.

Reply:
I will do this in a new PR. I started in this PR, but this actually touches a lot of points, so it might be easier to do it separately.

So in this PR lets go with the config we support currently.